tests/js/user-tests.js

```diff
@@ -150,107 +222,95 @@ module.exports = {
+    assertIsSameUser(app.currentUser(), firstUser);
+    let secondUser = await registerAndLogInEmailUser(app);
+    assertIsSameUser(app.currentUser(), secondUser); // the most recently logged in user is considered current
+    await secondUser.logOut();
+    assertIsSameUser(app.currentUser(), firstUser); // auto change back to another logged in user
+    await firstUser.logOut();
+    TestCase.assertNull(app.currentUser());
+  },
 
-  /* FIXME: do this with an email/pw user
-  testAll() {
+  async testAllWithEmail() {
     let app = new Realm.App(appConfig);
-    Object.keys(app.allUsers()).forEach(id => users[id].logOut()); // FIXME: we need to reset users for each test
+    await logOutExistingUsers(app);
 
-    const all = app.allUsers();
+    let all = app.allUsers();
     TestCase.assertArrayLength(Object.keys(all), 0, "Noone to begin with");
 
     let credentials = Realm.Credentials.anonymous();
-    return app.logIn(credentials).then(user1 => {
-      const all = app.allUsers();
-      TestCase.assertArrayLength(Object.keys(all), 1, "One user");
-      assertIsSameUser(all[user1.identity], user1);
-
-      return app.logIn(Realm.Credentials.anonymous()).then(user2 => {
-        let all = app.allUsers();
-        TestCase.assertArrayLength(Object.keys(all), 2, "Two users");
-        // NOTE: the list of users is in latest-first order.
-        assertIsSameUser(all[user2.identity], user2);
-        assertIsSameUser(all[user1.identity], user1);
-
-        user2.logOut();
-        all = app.allUsers();
-        TestCase.assertArrayLength(Object.keys(all), 1, "Back to one user");
-        assertIsSameUser(all[user1.identity], user1);
-
-        user1.logOut();
-        all = app.allUsers();
-        TestCase.assertArrayLength(Object.keys(all), 0, "All gone");
-      });
-    });
-  },*/
+    let user1 = await registerAndLogInEmailUser(app);
+    all = app.allUsers();
+    TestCase.assertArrayLength(Object.keys(all), 1, "One user");
+    assertIsSameUser(all[user1.identity], user1);
+    let user2 = await registerAndLogInEmailUser(app);
+    all = app.allUsers();
+    TestCase.assertArrayLength(Object.keys(all), 2, "Two users");
+    // NOTE: the list of users is in latest-first order.
+    assertIsSameUser(all[user2.identity], user2);
+    assertIsSameUser(all[user1.identity], user1);
+
+    await user2.logOut();
+    all = app.allUsers();
+    assertIsSameUser(all[user2.identity], user2);
+    assertIsSameUser(all[user1.identity], user1);
+    TestCase.assertFalse(user2.isLoggedIn);
+    TestCase.assertTrue(user1.isLoggedIn);
+    TestCase.assertArrayLength(Object.keys(all), 2, "still holds references to both users");
+
+    await user1.logOut();
+    all = app.allUsers();
+    TestCase.assertFalse(user1.isLoggedIn);
+    TestCase.assertFalse(user2.isLoggedIn);
+    TestCase.assertArrayLength(Object.keys(all), 2, "still holds references to both users"); // FIXME: is this actually expected?
```

Comment:
I think it is expected - for now. I agree it is counter-intuitive that anonymous users are not removed when they log out. 

See also https://github.com/realm/realm-object-store/issues/1031